File: app/src/main/java/com/classicsviewer/app/database/dao/WordDao.kt
package com.classicsviewer.app.database.dao

import androidx.room.Dao
import androidx.room.Query
import com.classicsviewer.app.database.entities.WordEntity

@Dao
interface WordDao {
    // Find lines containing words that match the lemma
    @Query("""
        SELECT DISTINCT w.book_id, w.line_number
        FROM words w
        INNER JOIN lemma_map lm ON w.word = lm.word_form
        WHERE lm.lemma = :lemma
        ORDER BY w.book_id, w.line_number
        LIMIT :limit
    """)
    suspend fun findLinesWithLemma(lemma: String, limit: Int): List<LineReference>
    
    // Find lines containing words that match the lemma, filtered by language
    @Query("""
        SELECT DISTINCT w.book_id, w.line_number
        FROM words w
        INNER JOIN lemma_map lm ON w.word = lm.word_form
        INNER JOIN books b ON w.book_id = b.id
        INNER JOIN works wk ON b.work_id = wk.id
        INNER JOIN authors a ON wk.author_id = a.id
        WHERE lm.lemma = :lemma AND a.language = :language
        ORDER BY w.book_id, w.line_number
        LIMIT :limit
    """)
    suspend fun findLinesWithLemmaByLanguage(lemma: String, language: String, limit: Int): List<LineReference>
    
    // Count occurrences of a lemma
    @Query("""
        SELECT COUNT(DISTINCT w.book_id || '-' || w.line_number)
        FROM words w
        INNER JOIN lemma_map lm ON w.word = lm.word_form
        WHERE lm.lemma = :lemma
    """)
    suspend fun countLinesWithLemma(lemma: String): Int
    
    // Count occurrences of a lemma, filtered by language
    @Query("""
        SELECT COUNT(DISTINCT w.book_id || '-' || w.line_number)
        FROM words w
        INNER JOIN lemma_map lm ON w.word = lm.word_form
        INNER JOIN books b ON w.book_id = b.id
        INNER JOIN works wk ON b.work_id = wk.id
        INNER JOIN authors a ON wk.author_id = a.id
        WHERE lm.lemma = :lemma AND a.language = :language
    """)
    suspend fun countLinesWithLemmaByLanguage(lemma: String, language: String): Int
    
    // Find lines with lemma and include word positions
    @Query("""
        SELECT DISTINCT line_info.book_id, line_info.line_number, line_info.sequence_number,
               GROUP_CONCAT(line_info.word || ':' || line_info.word_position) as word_positions
        FROM (
            SELECT w.book_id, w.line_number, w.sequence_number, w.word, w.word_position
            FROM words w
            INNER JOIN lemma_map lm ON w.word = lm.word_form
            WHERE lm.lemma = :lemma
        ) AS line_info
        GROUP BY line_info.book_id, line_info.line_number, line_info.sequence_number
        ORDER BY line_info.book_id, line_info.line_number, line_info.sequence_number
        LIMIT :limit
    """)
    suspend fun findLinesWithLemmaAndPositions(lemma: String, limit: Int): List<LineReferenceWithWords>
    
    // Find lines with lemma and include word positions, filtered by language
    @Query("""
        SELECT DISTINCT line_info.book_id, line_info.line_number, line_info.sequence_number,
               GROUP_CONCAT(line_info.word || ':' || line_info.word_position) as word_positions
        FROM (
            SELECT w.book_id, w.line_number, w.sequence_number, w.word, w.word_position
            FROM words w
            INNER JOIN lemma_map lm ON w.word = lm.word_form
            INNER JOIN books b ON w.book_id = b.id
            INNER JOIN works wk ON b.work_id = wk.id
            INNER JOIN authors a ON wk.author_id = a.id
            WHERE lm.lemma = :lemma AND a.language = :language
        ) AS line_info
        GROUP BY line_info.book_id, line_info.line_number, line_info.sequence_number
        ORDER BY line_info.book_id, line_info.line_number, line_info.sequence_number
        LIMIT :limit
    """)
    suspend fun findLinesWithLemmaAndPositionsByLanguage(lemma: String, language: String, limit: Int): List<LineReferenceWithWords>
}

data class LineReference(
    val book_id: String,
    val line_number: Int
)

data class LineReferenceWithWords(
    val book_id: String,
    val line_number: Int,
    val sequence_number: Int,
    val word_positions: String  // Format: "word1:pos1,word2:pos2,..."
)

File: ios/ClassicsViewer/Database/WordDAO.swift
import Foundation
import SQLite3

protocol WordDAOProtocol {
    func searchWords(query: String, bookId: String?, normalized: Bool) async throws -> [WordOccurrence]
    func getWordsAtPositions(bookId: String, lineNumber: Int, positions: [Int]) async throws -> [Word]
}

class WordDAO: WordDAOProtocol {
    // Use async database manager

    func searchWords(query: String, bookId: String?, normalized: Bool) async throws -> [WordOccurrence] {
        var queryString = """
            SELECT DISTINCT
                w.word as word,
                w.book_id,
                bk.label,
                COALESCE(wo.title_english, wo.title) as work_title,
                a.name as author_name,
                w.line_number,
                tl.line_text,
                GROUP_CONCAT(w.word_position) as positions,
                a.language
            FROM words w
            JOIN books bk ON w.book_id = bk.id
            JOIN works wo ON bk.work_id = wo.id
            JOIN authors a ON wo.author_id = a.id
            JOIN text_lines tl ON w.book_id = tl.book_id AND w.line_number = tl.line_number
            WHERE w.word LIKE ?
        """

        var parameters: [Any?] = ["%\(query)%"]

        if let bookId = bookId {
            queryString += " AND w.book_id = ?"
            parameters.append(bookId)
        }

        queryString += """
            GROUP BY w.book_id, w.line_number, w.word
            ORDER BY a.name, COALESCE(wo.title_english, wo.title), bk.book_number, w.line_number
            LIMIT 500
        """
        
        return try await DatabaseManagerAsync.shared.executeQuery(queryString, parameters: parameters) { [self] statement in
            wordOccurrenceFromStatement(statement)
        }
    }

    func getWordsAtPositions(bookId: String, lineNumber: Int, positions: [Int]) async throws -> [Word] {
        let positionPlaceholders = positions.map { _ in "?" }.joined(separator: ", ")

        let query = """
            SELECT id, word, book_id, line_number, sequence_number, word_position
            FROM words
            WHERE book_id = ? AND line_number = ? AND word_position IN (\(positionPlaceholders))
            ORDER BY word_position
        """

        var parameters: [Any?] = [bookId, lineNumber]
        parameters.append(contentsOf: positions)

        return try await DatabaseManagerAsync.shared.executeQuery(query, parameters: parameters) { [self] statement in
            wordFromStatement(statement)
        }
    }
    
    private func wordOccurrenceFromStatement(_ statement: OpaquePointer) -> WordOccurrence? {
        guard let wordCString = sqlite3_column_text(statement, 0),
              let bookIdCString = sqlite3_column_text(statement, 1),
              let workTitleCString = sqlite3_column_text(statement, 3),
              let authorNameCString = sqlite3_column_text(statement, 4),
              let lineTextCString = sqlite3_column_text(statement, 6),
              let positionsCString = sqlite3_column_text(statement, 7),
              let languageCString = sqlite3_column_text(statement, 8) else {
            return nil
        }

        let word = String(cString: wordCString)
        let bookId = String(cString: bookIdCString)
        let workTitle = String(cString: workTitleCString)
        let authorName = String(cString: authorNameCString)
        let lineNumber = Int(sqlite3_column_int(statement, 5))
        let lineText = String(cString: lineTextCString)
        let language = String(cString: languageCString)

        // Add book label to title if available
        var bookTitle = workTitle
        if let labelCString = sqlite3_column_text(statement, 2) {
            let label = String(cString: labelCString)
            bookTitle = "\(workTitle) - \(label)"
        }

        let positionsString = String(cString: positionsCString)
        let wordPositions = positionsString.split(separator: ",").compactMap { Int($0) }

        return WordOccurrence(
            word: word,
            bookId: bookId,
            bookTitle: bookTitle,
            authorName: authorName,
            lineNumber: lineNumber,
            lineText: lineText,
            wordPositions: wordPositions,
            language: language
        )
    }
    
    private func wordFromStatement(_ statement: OpaquePointer) -> Word? {
        let id = Int(sqlite3_column_int(statement, 0))
        
        guard let wordCString = sqlite3_column_text(statement, 1),
              let bookIdCString = sqlite3_column_text(statement, 2) else {
            return nil
        }
        
        let word = String(cString: wordCString)
        let bookId = String(cString: bookIdCString)
        let lineNumber = Int(sqlite3_column_int(statement, 3))
        let sequenceNumber = Int(sqlite3_column_int(statement, 4))
        let wordPosition = Int(sqlite3_column_int(statement, 5))
        
        return Word(
            id: id,
            word: word,
            bookId: bookId,
            lineNumber: lineNumber,
            sequenceNumber: sequenceNumber,
            wordPosition: wordPosition
        )
    }
}